File: P19/consultar_distancias/distancias.py
import mysql.connector
from mysql.connector import Error
import sys
import json

# Función para conectar a la base de datos
def conectar_bd():
    try:
        bd = mysql.connector.connect(
            host="localhost",
            user="sensor",
            password="1234",
            database="SENSORES"
        )
        return bd
    except Error as e:
        print("Error de conexión:", e)
        sys.exit(1)

# Función para realizar la búsqueda según el tipo especificado
def buscar_datos(busqueda, tipo_busqueda):
    try:
        conexion = conectar_bd()
        cursor = conexion.cursor(dictionary=True)

        if tipo_busqueda == "todos":
            consulta = "SELECT * FROM DISTANCIA"
            cursor.execute(consulta)
        elif tipo_busqueda == "distancia":
            consulta = "SELECT * FROM DISTANCIA WHERE CM = %s"
            cursor.execute(consulta, (busqueda,))
        elif tipo_busqueda == "fecha":
            consulta = "SELECT * FROM DISTANCIA WHERE FECHA = %s"
            cursor.execute(consulta, (busqueda,))
        elif tipo_busqueda == "hora":
            consulta = "SELECT * FROM DISTANCIA WHERE HORA = %s"
            cursor.execute(consulta, (busqueda,))
        elif tipo_busqueda == "count":
            consulta = "SELECT COUNT(*) AS total FROM DISTANCIA"
            cursor.execute(consulta)
            resultado = cursor.fetchone()
            total=str(resultado['total'])  # Devolver el resultado como texto
            return '<h3 style="color: green">Total de datos = '+total+'</h3>'  # Devolver el resultado como texto
        elif tipo_busqueda == "max":
            consulta = "SELECT MAX(CM) AS maximo FROM DISTANCIA"
            cursor.execute(consulta)
            resultado = cursor.fetchone()
            max= str(resultado['maximo'])  # Devolver el resultado como texto
            return '<h3 style="color: green">La distancia máxima es = '+max+'</h3>'  # Devolver el resultado como texto
        elif tipo_busqueda == "min":
            consulta = "SELECT MIN(CM) AS minimo FROM DISTANCIA"
            cursor.execute(consulta)
            resultado = cursor.fetchone()
            min=str(resultado['minimo'])
            return '<h3 style="color: green">La distancia mínima es = '+min+'</h3>'  # Devolver el resultado como texto
        elif tipo_busqueda == "avg":
            consulta = "SELECT AVG(CM) AS promedio FROM DISTANCIA"
            cursor.execute(consulta)
            resultado = cursor.fetchone()
            prom= str(resultado['promedio'])
            return '<h3 style="color: green">El promedio de distancias es = '+prom+'</h3>'  # Devolver el resultado como texto
        else:
            print("Tipo de búsqueda no válido")
            sys.exit(1)

        registros = cursor.fetchall()
        conexion.close()

        # Convertir los registros a formato JSON y devolver como texto
        return json.dumps(registros)

    except Error as e:
        print("Error en la consulta:", e)
        sys.exit(1)

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Uso: distancias.py <busqueda> <tipo_busqueda>")
        sys.exit(1)

    busqueda = sys.argv[1]
    tipo_busqueda = sys.argv[2]

    datos = buscar_datos(busqueda, tipo_busqueda)
    print(datos)  # Imprimir el resultado

File: P19/consultar_distancias/distancias.php
<!DOCTYPE html>
<html lang="es">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Distancia consultas</title>
    <style>
        table {
            width: 100%;
            border-collapse: collapse;
        }
        table, th, td {
            border: 1px solid black;
            padding: 8px;
            text-align: left;
        }
        th {
            background-color: #f4d03f;
        }
        td{
            background-color: #ebdef0 ;
        }
    </style>
</head>
<body>
    <h1 style="text-align: center; color: blue">Consulta de datos</h1>
    <form action="" method="post">
        <p style="text-align: center;">Buscar&nbsp;&nbsp;
            <input type="text" name="busqueda" size="20" maxlength="20" value="Todos">&nbsp;
            <select name="tipoBusqueda">
                <option value="todos" selected>Todo</option>
                <option value="distancia">Distancia</option>
                <option value="fecha">Fecha</option>
                <option value="hora">Hora</option>
                <option value="count">Total de datos</option>
                <option value="max">Distancia Máxima</option>
                <option value="min">Distancia Mínima</option>
                <option value="avg">Promedio de distancias</option>
            </select>
            <input type="submit" name="buscar" value="Buscar">
        </p>
    </form> 
    
    <?php
        if ($_SERVER["REQUEST_METHOD"] == "POST") {
            // Obtener los valores del formulario
            $busqueda = $_POST["busqueda"];
            $tipoBusqueda = $_POST["tipoBusqueda"];

            // Construir el comando para ejecutar el script Python
            $comando = "sudo python3 /var/www/html/consultar_distancias/distancias.py \"$busqueda\" \"$tipoBusqueda\"";

            // Ejecutar el comando y capturar la salida
            $salida = shell_exec($comando);

            // Mostrar la salida en una tabla si hay resultados
            if (!empty($salida)) {
                echo "<h2>Resultados:</h2>";
                if ($tipoBusqueda == "count" || $tipoBusqueda == "max" || $tipoBusqueda == "min" || $tipoBusqueda == "avg") {
                    // Si es una consulta que devuelve un solo valor
                    echo "<p>{$salida}</p>";
                } else {
                    // Si son múltiples registros
                    $registros = json_decode($salida, true); // Decodificar JSON a array asociativo
                    if (!empty($registros)) {
                        // Encabezados de la tabla
                        echo "<table>";
                        echo "<tr><th>DISTANCIA (CM)</th><th>FECHA</th><th>HORA</th></tr>";
                        // Filas de datos
                        foreach ($registros as $registro) {
                            $distancia = $registro['CM'];
                            $fecha = $registro['FECHA'];
                            $hora = $registro['HORA'];
                            echo "<tr><td>{$distancia}</td><td>{$fecha}</td><td>{$hora}</td></tr>";
                        }
                        echo "</table>";
                    } else {
                        echo '<h3 style="color: red">No se encontraron resultados.</h3>';
                    }
                }
            } else {
                echo "<p>No se obtuvieron resultados.</p>";
            }
        }
    ?>
</body>
</html>
